### ISSUE TITLE
Coroutine usage in heavy responses from servers

### ISSUE THREAD

**Issue by asmodeus812** [neovim/neovim#26924]
### Problem

The issue is related to the fact that nvim blocks while handlers are transforming data on lsp server response. Most noticeable in things like lsp references (in my case a test with 2000 entries in references, takes up to 10 seconds to show up in the qf list, after the server responds with the results - the server actually returns the results faster (in about 2-3 secs) than in takes nvim to process them in the qf list), which could, in some cases receive a huge amount of entries, other lsp calls could too (or filter over workspace symbols, with short or widely matching query). 

The issue with lsp references is that once the lsp handler is called, editor blocks while results are being transformed. That processing should be done in a coroutine, where control should be yielded back to the main thread, the coroutine should be only have a fixed about of time to work/do it's thing, and give control back, to unblock the editor.

### Expected behavior

The editor should not be blocking while the handler is processing the results from lsp requests. There could be a general solution or api where we are also allowed to easily customize and transform the results in handlers and the coroutine action happens transparently to the user / api consumer.

**Comment by wookayin** [comment]
Just to make sure, in your config/usage LSP request-response is done asynchronously rather than synchronously, right? If you have any benchmark/profiling, or minimal reproduction to demonstrate the performance issue you're having, that'd be very helpful to make sure we are all on the same page. `vim.lsp.util.locations_to_items()` might be the synchronous bottleneck but I'm not quite sure this will be *that* slow.

I can also notice a little short freezing as soon as the response handler callback is called, when the # of items in a response to a `textDocument/references` request is >10,000, but this is kind of bearable (0.1~0.2 seconds) -- not as annoying as yours (10 seconds).

**Comment by asmodeus812** [comment]
Yes, i have been testing this with a default configuration, the request is as always async, and after the request is done, i can browse through the editor for about 2-3 seconds, until the response comes in, when it does nvim locks up until the qflist is opened with the results.

**Comment by asmodeus812** [comment]
Here is some brief information, after calling references, the response comes in 3 seconds, during which the editor is NOT blocked (as expected)

[results.txt](https://github.com/neovim/neovim/files/13850741/results.txt) - results table as returned by the server.

```
spent in buf_request: 0 sec
spent in locations_to_items: 13 sec
spent in after qf population: 0 sec
```

**Comment by wookayin** [comment]
The reason why `locations_to_items()` is slow is: because it *reads* the line for each location item *eagerly* (if the file/buffer is not loaded, it loads a buffer and then read it):

https://github.com/neovim/neovim/blob/v0.9.5/runtime/lua/vim/lsp/util.lua#L1822
https://github.com/neovim/neovim/blob/v0.9.5/runtime/lua/vim/lsp/util.lua#L237-L317

which is all synchronous, not quite efficient. This also has something to do with disk I/O speed. On Java, things would be much worse because reading would need to deal with extracting JAR (zip) files.... This should be improved by deferring `get_lines`, to be lazily read in an asynchronous context.

**Comment by lewis6991** [comment]
Related https://github.com/neovim/neovim/issues/20783

**Comment by lewis6991** [comment]
`setqflist` allows adding to an existing list so we could break up `locations_to_items` over several event loop iterations. We already do this for semantic tokens.

**Comment by justinmk** [comment]
> `setqflist` allows adding to an existing list so we could break up `locations_to_items` over several event loop iterations

Somewhat related? https://github.com/neovim/neovim/pull/30868
